### DISCUSSION TITLE
Specifying boot parameters from UEFI for Orin

### DISCUSSION THREAD

**Discussion by formidable01** [OE4T/meta-tegra#1191] [Q&A]
Hi,  I am new to the Nvidia yocto world and have what is probably a basic question.  I am trying to integrate use of [RAUC ](https://rauc.readthedocs.io/en/latest/updating.html) as an update solution for the Orin system.  The use of UEFI in the Orin system seems to complicate things because the kernel needs to be booted with some RAUC-specific parameters to make the process work (or the[ efimanager ](https://rauc.readthedocs.io/en/latest/integration.html?highlight=efi#efi) needs to be run with specified parameters to declare multiple EFI images and their locations along with an identifying label) My problem is figuring out where and how to change the yocto generated boot to incorporate this.  Any clues on where and how to do this would be appreciated.

**Comment by madisongh** [comment]
Not a basic question at all.  The Jetsons are new to the UEFI world, and it isn't clear to me, at least, exactly how to go about this, based on the current state of NVIDIA's implementation.  It's at least a bit different from x86 EFI, so the process RAUC has for EFI integration probably won't work.  For one thing, NVIDIA's has an A/B boot chain implementation in addition to the normal boot order mechanism that you see on x86 systems. The `L4TLauncher` EFI app that is used to launch the OS is also a bit funky, implementing some of the features carried forward from their older `cboot` bootloader, which I'm not sure was a good idea.  Overall, the boot chain is just more complicated, since the firmware loaded during the early boot stages has to be compatible with the kernel - something that isn't typically an issue with x86 systems.

As far as getting RAUC to identify the correct boot slot, I could see a couple of approaches.  One would be to modify `L4TLauncher` in the UEFI sources to automatically insert the info into the kernel command line, and configure RAUC for "custom" bootloader integration (similar to the approach I took with cboot in older L4T releases).  Another would be to patch RAUC's EFI integration to handle the differences in the Jetson implementation - mainly dealing with the `BootChainFw` set of EFI variables it uses.  I haven't had a chance to look into this in detail, so I'm not sure which would be easier or better.  And I would expect to see some further changes coming in NVIDIA's UEFI implementation, so be warned that whatever you might implement now might have to change later.

**Reply by formidable01** [reply]
I think I will look into the `L4TLauncher `approach but I wonder if it is possible to generate an extlinux.conf file to set things up correctly specifically for RAUC instead of patching `L4TLauncher `or is it the case that extlinux.conf is not used by `L4TLauncher`?

**Reply by madisongh** [reply]
That's not a bad idea, actually.  In fact, I believe that's the approach that NVIDIA uses in stock L4T for their A/B support - they write different kernel args into the extlinux.conf file on the A and B partitions.

If you're free to modify the rootfs partition as part of the update process, you could have a script update the extlinux.conf file with the correct slot information.  If you need to be compatible with some form of delta update system, though, that might not be so good, since some of them don't work, or don't work as well, if the rootfs has been modified at all from what was generated at build time. You could split out the `/boot` subtree to a separate partition to mitigate that, if RAUC provides a way to handle multi-partition updates.  (A less-desirable approach would be to include full A and B rootfs images in your update package, and have RAUC install one or the other, but that seems like overkill for handling a one-line difference in one file.)

Also note that with secure boot enabled, the extlinux.conf file has to be signed, so you may need to pre-generate A and B copies of that file at build time, sign them, and the install the correct one as part of your postinstall script, rather than just modifying the file on the fly.

**Reply by formidable01** [reply]
I have spent the past few days trying to look into this.  First off, I do not see that the new UEFI-based Nvidia solution even uses extlinux.conf.  When you look at the /boot directory on the root filesystem - it is unpopulated. I tried to mount other sectors to look for it and can't see it anywhere.  The extlinux.conf that's in the source tree has no effect and does not even appear to be used to generate the UEFI menus or the boot command lines. Instead, this is all driven from the contents of NvidiaConfigHii.vfr file (and responses to the data pulled from that).  So, I need to understand that more and found this: https://github.com/Kostr/UEFI-Lessons which seems rather useful.  But without getting into that too far I started looking into maybe using the L4TConfiguration device tree overlay to pass kernel boot arguments since that device tree has an entry called KernelCommandLine.  Investigating that further resulted in a yet another conundrum.  KernelCommandLine is defined as unicode (made up of an array of uint16) but the device tree syntax and compiler does not have an obvious way of assigning a unicode string in a dts file.  So mostly I am venting and sharing.  Big picture-wise we are looking at using RAUC as an update system over the A/B stuff that Nvidia has now built in. So, we would have a hierarchy of RAUC-A which would have an Nvidia A/B system beneath it and RAUC-B which would have yet another Nvidia A/B system beneath it.  Right now, the reason for this is that Nvidia hasn't yet released its A/B support for Orin (but has, it seems, for Xavier).  It is apparently always coming in "the next release".  Even then, this approach could offer us some higher reliability and flexibility.

**Reply by madisongh** [reply]
>First off, I do not see that the new UEFI-based Nvidia solution even uses extlinux.conf. When you look at the /boot directory on the root filesystem - it is unpopulated.

The L4TLauncher EFI app does look for it, and falls back to using the Android-style kernel/kernel-dtb partitions if it's not present.  We just don't enable it in the builds by default.  Set `UBOOT_EXTLINUX = "1"` in your build configuration to enable.

**Reply by formidable01** [reply]
Sorry to be a bother here but when I set UBOOT_EXTLINUX = "1" I do indeed see the /boot directory populated but the boot behavior is identical as before and the contents of extlinux.conf are ignored

**Reply by madisongh** [reply]
Hmm.  I was just testing this yesterday and it was working... how are you determining that it's being ignored?  You should see some messages on the console from L4TLauncher during boot about it trying to locate the extlinux.conf file.

**Reply by formidable01** [reply]
I don't see the messaging you indicated. Perhaps I put UBOOT_EXTLINUX = "1" in the wrong place. I put it in meta-tegra/conf/machine/jetson-agx-orin-devkit.conf

**Reply by formidable01** [reply]
A little more info: The boot load reads:

L4TLauncher: Attempting GRUB Boot
L4TLauncher: Attempting Direct Boot
ProcessExtLinuxConfig: Unable to find partition info
L4TLauncher: Unable to process extlinux config: Not Found
L4TLauncher: Attempting Kernel Boot
EFI stub: Booting Linux Kernel...
EFI stub: Using DTB from configuration table

So not finding extlinux.conf maybe because the partition is not present?

**Reply by kekiefer** [reply]
Last night I encountered a similar issue, and found that if the rootfs ext4 partition has the flag `metadata_csum_seed`, then edk2 won't read the extlinux from there. You can check for the flag: `dumpe2fs /dev/<rootfs_device> | grep features`, and remove it to see: `tune2fs -O ^metadata_csum_seed /dev/<rootfs_device>`.

**Reply by formidable01** [reply]
So my additional complication is that I am building my image as a core-image-minimal so I do not have run time access to `dumpe2fs` or `tune2fs` so I would have to figure out how to generate the rootfs without that flag

**Reply by dwalkes** [reply]
https://github.com/OE4T/meta-tegra/commit/dddf1308a1485b17ee15b8e13e167f6641ab4521 is probably relevant

**Reply by formidable01** [reply]
Hmmm. So I made the modification suggested by the relevant commit above but I still get the exact same behavior which seems odd to me.

```
L4TLauncher: Attempting Direct Boot
ProcessExtLinuxConfig: Unable to find partition info
L4TLauncher: Unable to process extlinux config: Not Found
L4TLauncher: Attempting Kernel Boot
EFI stub: Booting Linux Kernel...
```

**Reply by madisongh** [reply]
Looking at the UEFI sources, that second message is emitted [here](https://github.com/NVIDIA/edk2-nvidia/blob/c647e59d6032aa721989e0824579de192b925277/Silicon/NVIDIA/Application/L4TLauncher/L4TLauncher.c#L1027-L1031), where it's looking for a partition named `APP` (that's ROOTFS_BASE_NAME). Have you modified your flash partition names?

**Reply by formidable01** [reply]
In fact, that's exactly what I did.  Since RAUC has 2 root filesystems, I changed the names to APP.0 and APP.1 To make this work I will have to change the code you identified to either look for those names or look for a name that starts with APP, I guess

**Reply by emrodre** [reply]
Hi @formidable01,

Were you able to integrate RAUC on the Orin?

I've not found any info on this and would love to know if I'm at a dead end.

Thanks!

**Reply by joekale-pp** [reply]
@emrodre, There are a couple people (including myself) who have used RAUC on JP5 and have documented some of our efforts on the OE4T Gitter chats.

**Reply by emrodre** [reply]
Thanks @joekale-pp! Will check the conversation there.
